lib/providers/ipmeta_provider_maxmind_v2.c

```diff
@@ -0,0 +1,870 @@
+    if ((state->locations_file = malloc(
+           strlen(directory) + 1 + strlen(LOCATIONS_FILE_NAME) + 1)) == NULL) {
+      ipmeta_log(__func__, "could not malloc location file string");
+      return -1;
+    }
+
+    if ((state->blocks_file = malloc(strlen(directory) + 1 +
+                                     strlen(BLOCKS_FILE_NAME) + 1)) == NULL) {
+      ipmeta_log(__func__, "could not malloc blocks file string");
+      return -1;
+    }
+
+    /** @todo make this check for both .gz and non-.gz files */
+
+    ptr = stpncpy(state->locations_file, directory, strlen(directory));
+    *ptr++ = '/';
+    /* last copy needs a +1 to get the terminating nul. d'oh */
+    ptr = stpncpy(ptr, LOCATIONS_FILE_NAME, strlen(LOCATIONS_FILE_NAME) + 1);
+
+    ptr = stpncpy(state->blocks_file, directory, strlen(directory));
+    *ptr++ = '/';
+    ptr = stpncpy(ptr, BLOCKS_FILE_NAME, strlen(BLOCKS_FILE_NAME) + 1);
+  }
+
+  if (state->locations_file == NULL || state->blocks_file == NULL) {
+    fprintf(stderr, "ERROR: %s requires either '-d' or both '-b' and '-l'\n",
+            provider->name);
+    usage(provider);
+    return -1;
+  }
+  return 0;
+}
+
+/* Parse a maxmind_v2 location cell */
+static void parse_maxmind_v2_location_cell(void *s, size_t i, void *data)
+{
+  ipmeta_provider_t *provider = (ipmeta_provider_t *)data;
+  ipmeta_provider_maxmind_v2_state_t *state = STATE(provider);
+  ipmeta_record_t *tmp = &(state->tmp_record);
+  char *tok = (char *)s;
+  char *end;
+  /* skip the first line */
+  if (state->current_line < HEADER_ROW_CNT) {
+    return;
+  }
+
+  switch (state->current_column) {
+  case LOCATION_COL_ID:
+    /* init this record */
+    tmp->id = strtol(tok, &end, 10);
+    if (end == tok || *end != '\0' || errno == ERANGE) {
+      ipmeta_log(__func__, "Invalid ID Value (%s)", tok);
+      state->parser.status = CSV_EUSER;
+      return;
+    }
+    break;
+
+  case LOCATION_COL_LOCALECODE:
+    /* country code */
+    if (tok == NULL || strlen(tok) != 2) {
+      ipmeta_log(__func__, "Invalid Locale Code (%s)", tok);
+      state->parser.status = CSV_EUSER;
+      return;
+    }
+    memcpy(tmp->locale_code, tok, 2);
+    break;
+
+  case LOCATION_COL_CONTINENTCODE:
+    /* continent/region code */
+    if (tok == NULL || strlen(tok) != 2) {
+      ipmeta_log(__func__, "Invalid  Continent Code (%s)", tok);
+      state->parser.status = CSV_EUSER;
+      return;
+    }
+    if (tok[0] == '-' && tok[1] == '-') {
+      tok[0] = '?';
+      tok[1] = '?';
+    }
+    memcpy(tmp->continent_code, tok, 2);
+    break;
+
+  case LOCATION_COL_REGION:
+    /* region string */
+    if (tok != NULL) {
+      if ((tmp->region = strdup(tok)) == NULL) {
+        ipmeta_log(__func__, "Region code copy failed (%s)", tok);
+        state->parser.status = CSV_EUSER;
+        return;
+      }
+    }
+    break;
+
+  case LOCATION_COL_CC:
+    /* country code */
+    /* Generates an error for cases where CC = NULL, so I commented it */
```

Comment:
there's no need to leave this type of code if it is wrong or broken. just delete it.
but in this case i think you might want to just change the check to something like `if (tok != NULL && strlen(tok) != 2)` -- that way you get an error for country codes that are set, but not 2 characters.